### DISCUSSION TITLE
What's the recommended way to check a type's value?

### DISCUSSION THREAD

**Discussion by nbrouwers** [eclipse-langium/langium#1314] [Q&A]
Hi all,

We are developing a DSL for the domain of Domain Driven Design. Amongst others, it contains Entities that have Attributes. Attributes have a name and type property. See below for a small example:

```
    Aggregate ItemBarcode {
        Entity ItemBarcode {
            - int id
            - Barcode barcode
        }
        ValueObject Barcode {
            string value
        }
   }
```

The attribute's type can be a TypeReference, e.g. referencing ValueObject _Barcode_, or a primitive/basic type, e.g. like assigning it a type like _int_.

For the development of a code generator for this DSL, we need to check the value of the Attribute's type. Specially, we'd like to do something like this:

```
    generateType(type:unknown) : Generated {
        let content : Generated;

        if (isTypeReference(type)) {
           // Fill content here 
        } else if (isType_Timestamp(type)) {
            content = 'Timestamp';
        } else if (isType_Integer(type)) {
            content = 'int';
        } else if (isType_String(type)) {
            content = 'string';
        } else {
            content = 'string'
        }
        return content
    } 
}
```

The problem is that the isType_<BasicType> variants from the generated ast.ts never evaluate to true.
We suspect this has to do with _BasicType_ not being considered as a specific type of _Type_, maybe since _BasicType_ is a Langium _type_ and not a parser rule?

So the question: could someone recommend a robust way to check on the value of a _BasicType_?

For reference, the relevant grammar is the following:
```
Entity infers Entity:
    "Entity" name=ID  
    '{' 
       ('-' attributes+=Attribute)* 
    "}"
;

Attribute infers Attribute:
    ^type=Type name=ID (isKey='key')?
;

Type infers Type:
    BasicType | TypeReference
;

TypeReference infers TypeReference:
    ref=[ValueObject:ID]
;

type BasicType =
      'string' 
    | 'int' 
    | 'double' 
    | 'float'
    | 'long' 
    | 'boolean' 
    | 'Date' 
    | 'DateTime' 
    | 'Timestamp' 
    | 'BigDecimal' 
    | 'BigInteger' 
;

BasicType returns BasicType:
      Type_String
    | Type_Integer 
    | Type_Double 
    | Type_Float
    | Type_Long 
    | Type_Boolean 
    | Type_Date 
    | Type_DateTime 
    | Type_Timestamp 
    | Type_BigDecimal 
    | Type_BigInteger
;

Type_String returns string: "string";
Type_Integer returns string: "int";
Type_Double returns string: "double";
Type_Float returns string: "float";
Type_Long returns string: "long";
Type_Boolean returns string: "boolean";
Type_Date returns string: "Date";
Type_DateTime returns string: "DateTime";
Type_Timestamp returns string: "Timestamp";
Type_BigDecimal returns string: "BigDecimal";
Type_BigInteger returns string: "BigInteger";
```

**Comment by msujew** [comment]
Hey @nbrouwers,

looking at the generated `is<type>` checking code, everything looks fine:

```ts
export function isType_BigDecimal(item: unknown): item is Type_BigDecimal {
    return item === 'BigDecimal';
}
```

Can you confirm that this is generated for your language as well? Can you also confirm (maybe via debugging or similar) that the values you pass in there are actually what you think you're passing in there? Given that the generated checking function is pretty simple, I assume there's something else going on.

**Comment by nbrouwers** [comment]
Hi @msujew,

thanks for your quick reply.
Much appreciated!

The generateType function is called from the generateAttribute function as shown below:
```typescript
generateAttribute(gen) {
    return toNode `${gen.name} ${this.generateType(gen.type)}`;
}
```
In this context, the most generic type to pass is _Type_, generalizing both _TypeReference_ and _BasicType_.
When it enters the generateType method, it indeed is passed as a _Type_.

Note that the check _isTypeReference_ in the beginning works as expected; in case the _type_ parameter is of type _TypeReference_, the expression evaluates to true.

And yes, I can confirm that these check functions are generated for all _BasicType_ types.

```typescript
export type BasicType = 'BigDecimal' | 'BigInteger' | 'Date' | 'DateTime' | 'Timestamp' | 'boolean' | 'double' | 'float' | 'int' | 'long' | 'string';
```

```typescript
export type Type_BigDecimal = 'BigDecimal';

export function isType_BigDecimal(item: unknown): item is Type_BigDecimal {
    return item === 'BigDecimal';
}

export type Type_BigInteger = 'BigInteger';

export function isType_BigInteger(item: unknown): item is Type_BigInteger {
    return item === 'BigInteger';
}

export type Type_Boolean = 'boolean';

export function isType_Boolean(item: unknown): item is Type_Boolean {
    return item === 'boolean';
}

export type Type_Date = 'Date';

export function isType_Date(item: unknown): item is Type_Date {
    return item === 'Date';
}

export type Type_DateTime = 'DateTime';

export function isType_DateTime(item: unknown): item is Type_DateTime {
    return item === 'DateTime';
}

export type Type_Double = 'double';

export function isType_Double(item: unknown): item is Type_Double {
    return item === 'double';
}

export type Type_Float = 'float';

export function isType_Float(item: unknown): item is Type_Float {
    return item === 'float';
}

export type Type_Integer = 'int';

export function isType_Integer(item: unknown): item is Type_Integer {
    return item === 'int';
}

export type Type_Long = 'long';

export function isType_Long(item: unknown): item is Type_Long {
    return item === 'long';
}

export type Type_String = 'string';

export function isType_String(item: unknown): item is Type_String {
    return item === 'string';
}

export type Type_Timestamp = 'Timestamp';

export function isType_Timestamp(item: unknown): item is Type_Timestamp {
    return item === 'Timestamp';
}
```

**Reply by nbrouwers** [reply]
Btw, one other possibly important observation:

- In case the type of the Attribute is a _TypeReference_, the _type_ parameter as it enters the _generateType_ function is indeed of type _TypeReference_.
- However, in case the type is (supposed to be) a _BasicType_, the _type_ parameter as it enters the _generateType_ function is of type _Type_ (instead of BasicType).

The latter is of course suspicious. Why is it passed as the more generic _Type_?

**Reply by nbrouwers** [reply]
Btw. one other possibly important observation:

_Type_ seems to be aliased by _TypeReference_, where we otherwise would expect it to be a generic type specializing both _TypeReference_ and _BasicType_.

```ts
export type Type = TypeReference;

export const Type = 'Type';
```

**Reply by msujew** [reply]
The issue is probably that `Type` is inferred as an AST type, while all your `BasicType` types are actually just of type `string`. This probably messes with the generator and parser code in some way. I would recommend to refactor it:

```
Type infers Type:
    SimpleType | TypeReference
;

SimpleType: value=BasicType;
```

**Reply by nbrouwers** [reply]
I did the refactoring as suggested (but with different naming), and the issue is the same. 

The check function now receives indeed a _SimpleType_ as _item_ parameter, but the implementation is not correct. Instead of checking the _item_ itself to be equal to the _SimpleType_ value (e.g. _int_), it should check against _item_'s _value_ property.

So in this case, the correct implementation would be:
```ts
export function isType_Integer(item: unknown): item is Type_Integer {
    return item.value === 'int';
}
```

Despite this, your refactoring suggestion does now allow us to implement a simple check ourselves, since it's now a _SimpleType_ entering the _generateType_ function, which has an explicit _value_ property with the _BasicType_ as string representation.

Still, I think the AST generation can be improved to support this pattern. How to file this as a feature request?

Thank you so much for your help!

**Reply by msujew** [reply]
> So in this case, the correct implementation would be:

No, the generated code is still correct. You just need to pass in `item.value` yourself into `isType_Integer`.

> since it's now a SimpleType entering the generateType function, which has an explicit value property with the BasicType as string representation.

Yes, exactly, that was the purpose. You now just need to put an additional `isSimpleType` in there and you're done :)

> Still, I think the AST generation can be improved to support this pattern. How to file this as a feature request?

You can go ahead and file an issue for the generated

```ts
export type Type = TypeReference;

export const Type = 'Type';
```

stuff. It's supposed to be:

```ts
export type Type = BasicType | TypeReference;

export const Type = 'Type';
```
